edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs

```diff
@@ -128,9 +124,9 @@ static async Task<int> MainAsync(IConfigurationRoot configuration)
             {
                 try
                 {
-                    await Task.WhenAll(mqttBrokerProtocolHead.StartAsync(), configDownloadTask);
-                    await edgeHubProtocolHead.StartAsync();
-                    await Task.WhenAny(cts.Token.WhenCanceled(), renewal.Token.WhenCanceled(), configUpdaterStartupFailed.Task);
+                    await configDownloadTask;
+                    await Task.WhenAll(mqttBrokerProtocolHead.StartAsync(), edgeHubProtocolHead.StartAsync());
```

Comment:
Do we want the MqttBrokerProtocolHead to start before the EdgeHubProtocolHeads?
@vipeller ?

Reply:
Yes, I thing there were some ordering requirement here, that is how the code evolved this way (from just listing all protocol heads and start them at once).
One restriction is that authentication must be available later then mqtt(broker), but that is ensured inside 'GetMqttBrokerProtocolHeadAsync()'
I do not remember if that was bad that the mqtt related protocol heads start up later than the rest, but let's assume that the previous implementation (start mqtt first and then the rest) happened for reason and let's keep that order.

Reply:
Ohh I think there is another problem: let's say that this is a nested scenario and the first time edge starts up, so there is no cached config (or the config is missing for other reason).
Now what will happen is that the config falls back to pulling it from upstream. However, if the broker is enabled, then the mqttbrokerprotocolhead needs to be started, otherwise the code which would be able to execute request upstream (pulling the twin = pulling the config) will not able to complete. So in case of mqtt turned on, the two awaits above are basically a dead lock: configDownloadTask is waiting for the broker protocol head and the broker protocol head (start-up) is waiting for the configDownload.